db/system_keyspace.cc

```diff
@@ -1838,6 +1839,101 @@ class nodetool_status_table : public memtable_filling_virtual_table {
+private:
+    service::storage_service& _ss;
+public:
+    describe_ring_table(service::storage_service& ss)
+            : streaming_virtual_table(build_schema())
+            , _ss(ss)
+    {
+        _shard_aware = true;
+    }
+
+    static schema_ptr build_schema() {
+        auto id = generate_legacy_id(NAME, "describe_ring");
+        return schema_builder(NAME, "describe_ring", std::make_optional(id))
+            .with_column("keyspace_name", utf8_type, column_kind::partition_key)
+            .with_column("start_token", utf8_type, column_kind::clustering_key)
+            .with_column("endpoint", inet_addr_type, column_kind::clustering_key)
+            .with_column("end_token", utf8_type)
+            .with_column("dc", utf8_type)
+            .with_column("rack", utf8_type)
+            .with_version(generate_schema_version(id))
+            .build();
+    }
+
+    dht::decorated_key make_partition_key(const sstring& name) {
+        return dht::decorate_key(*_s, partition_key::from_single_value(*_s, data_value(name).serialize_nonnull()));
+    }
+
+    clustering_key make_clustering_key(sstring start_token, gms::inet_address host) {
+        return clustering_key::from_exploded(*_s, {
+            data_value(start_token).serialize_nonnull(),
+            data_value(host).serialize_nonnull()
+        });
+    }
+
+    struct endpoint_details_cmp {
+        bool operator()(const dht::endpoint_details& l, const dht::endpoint_details& r) const {
+            return inet_addr_type->less(
+                data_value(gms::inet_address(l._host)).serialize_nonnull(),
+                data_value(gms::inet_address(r._host)).serialize_nonnull());
+        }
+    };
+
+    future<> execute(reader_permit permit, result_collector& result, const query_restrictions& qr) override {
+        struct decorated_keyspace_name {
+            sstring name;
+            dht::decorated_key key;
+        };
+
+        auto keyspace_names = boost::copy_range<std::vector<decorated_keyspace_name>>(
+                _db->get_keyspaces() | boost::adaptors::transformed([this] (auto&& e) {
+                    return decorated_keyspace_name{e.first, make_partition_key(e.first)};
+        }));
+
+        boost::sort(keyspace_names, [less = dht::ring_position_less_comparator(*_s)]
+                (const decorated_keyspace_name& l, const decorated_keyspace_name& r) {
+            return less(l.key, r.key);
+        });
+
+        for (const decorated_keyspace_name& e : keyspace_names) {
+            auto&& dk = e.key;
+            if (!this_shard_owns(dk) || !contains_key(qr.partition_range(), dk)) {
+                continue;
+            }
+
+            std::vector<dht::token_range_endpoints> ranges;
+            try {
+                ranges = _ss.describe_ring(e.name);
+            } catch (std::runtime_error& ex) {
+               slogger.warn("Could not describe ring for keyspace {}: {}", e.name, ex.what());
```

Comment:
I think we should fail the read rather than return empty results on generic error. If you're covering for concurrent delete, then you can re-check for existence here to be safe.